sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/PreparedStatementAndResultSetCheck.kt

```diff
@@ -0,0 +1,87 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.kotlin.checks
+
+import org.jetbrains.kotlin.psi.KtCallExpression
+import org.jetbrains.kotlin.psi.KtExpression
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.calls.model.ResolvedCall
+import org.jetbrains.kotlin.resolve.calls.util.getCall
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.CallAbstractCheck
+import org.sonarsource.kotlin.api.FunMatcher
+import org.sonarsource.kotlin.api.determineTypeAsString
+import org.sonarsource.kotlin.api.predictReceiverExpression
+import org.sonarsource.kotlin.api.predictRuntimeIntValue
+import org.sonarsource.kotlin.api.predictRuntimeStringValue
+import org.sonarsource.kotlin.api.predictRuntimeValueExpression
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+
+private val PREPARE_STATEMENT = FunMatcher(qualifier = "java.sql.Connection", name = "prepareStatement")
+
+private val PREPARE_STATEMENT_SET = FunMatcher(qualifier = "java.sql.PreparedStatement", nameRegex = """^set""".toRegex())
+
+private val RESULT_SET_GET = FunMatcher(qualifier = "java.sql.ResultSet", nameRegex = """^get""".toRegex())
+
+@Rule(key = "S2695")
+class PreparedStatementAndResultSetCheck : CallAbstractCheck() {
+
+    override val functionsToVisit = listOf(PREPARE_STATEMENT_SET, RESULT_SET_GET)
+
+    override fun visitFunctionCall(
+        callExpression: KtCallExpression,
+        resolvedCall: ResolvedCall<*>,
+        kotlinFileContext: KotlinFileContext,
+    ) {
+        val bindingContext = kotlinFileContext.bindingContext
+
+        val firstArgument = callExpression.valueArguments[0].getArgumentExpression() ?: return
+        val receiver = callExpression.predictReceiverExpression(bindingContext) ?: return
+
+        val firstArgumentValue = firstArgument.predictRuntimeIntValue(bindingContext) ?: return
+        val isResultSet = receiver.determineTypeAsString(bindingContext).equals("java.sql.ResultSet")
+
+        if (isResultSet && firstArgumentValue == 0) {
+            kotlinFileContext.reportIssue(firstArgument, "ResultSet indices start at 1.")
+        } else if (!isResultSet) {
+            if (firstArgumentValue == 0) {
+                kotlinFileContext.reportIssue(firstArgument, "PreparedStatement indices start at 1.")
+            } else {
+                val numberOfParameters = getNumberOfParameters(bindingContext, receiver, firstArgumentValue)
+                if (firstArgumentValue > numberOfParameters) {
+                    kotlinFileContext.reportIssue(
+                        firstArgument,
+                        "This \"PreparedStatement\" " + (if (numberOfParameters == 0) "has no" else "only has $numberOfParameters") + " parameters."
+                    )
+                }
+            }
+        }
+    }
+}
+
+private fun getNumberOfParameters(bindingContext: BindingContext, receiver: KtExpression, defaultValue: Int) =
```

Comment:
You can return nullable here and drop `defaultValue`. Then on the call site, you can do 
```
getNumberOfParameters(bindingContext, receiver)?.let {...}
```